docs/advisory.rst

```diff
@@ -0,0 +1,111 @@
+.. _advisory:
+
+Advisory Locks
+==============
+
+Sometimes applications need to perform custom locking that cannot be easily expressed as
+row or table locks. The `pglock.advisory` decorator and context manager solves this by using
+`Postgres advisory locks <https://www.postgresql.org/docs/current/explicit-locking.html#ADVISORY-LOCKS>`__.
+
+Basic Examples
+--------------
+
+`pglock.advisory` can be used to acquire an advisory lock. By default, this means:
+
+1. It will block until the lock is available.
+2. Once acquired, no other code that requests the same lock will be able to run.
+
+For example, below we've ensured there's only one instance of ``my_exclusive_function``
+running at a time:
+
+.. code-block:: python
+
+    import pglock
+
+    @pglock.advisory("my_module.my_exclusive_function")
+    def my_exclusive_function():
+        # All other calls of my_exclusive_function will wait for this one to finish
+
+.. tip::
+
+    When used as a decorator, the lock ID isn't required
+    and will default to ``<module_name>.<function_name>``.
+
+When creating an advisory lock, remember that the lock ID is a global name across the entire database. Be sure
+to choose meaningful names, ideally with namespaces, when serializing code with `pglock.advisory`.
+
+Configuring Lock Wait Time
+--------------------------
+
+By default, `pglock.advisory` will wait forever until the lock can be acquired. There are two ways to configure
+this behavior:
+
+1. Use ``wait=False`` to avoid waiting for the lock.
+2. Use the ``timeout`` parameter to wait a configurable amount of time.
+
+For example, below we use `pglock.advisory` as a context manager without waiting:
+
+.. code-block:: python
+
+    with pglock.advisory("my_lock_id", wait=False) as acquired:
+        if not acquired:
+            # Do stuff if the lock cannot be acquired
+        else:
+            # The lock was acquired
+
+As shown above, the context manager returns a flag to indicate if the lock was successfully acquired.
+The same holds true for using a timeout:
+
+.. code-block:: python
+
+    with pglock.advisory("my_lock_id", timeout=pglock.timedelta(seconds=2)) as acquired:
+        ...
+
+Above we've ensured that we wait at most two seconds for the lock to be acquired.
+`pglock.timedelta` is simply an alias for Python's ``timedelta`` object for convenience.
+
+.. note::
+
+    You might be asking, why not just use ``timeout=pglock.timedelta(seconds=0)`` instead
+    of ``wait=False``? These arguments are separate because they implement different
+    strategies under the hood. ``wait=False`` uses a different Postgres function
+    to acquire the lock. It's recommended to use this when waiting isn't required.
```

Comment:
I was originally against it because I thought it was lengthy to have to use the ``timeout`` argument and supply a timedelta every time. What I think I can do instead is allow ``timeout`` to take a number in seconds or a timedelta object.

I.e. `pglock.advisory(timeout=0)` will perform NOWAIT, and I can update `pglock.model` to have the same behavior